File: php/src/google_chat.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function google_chat_parse_event(array $payload): array
{
    $message=$payload['message']??$payload;$sender=$message['sender']??[];$space=$payload['space']??[];$user=$sender['name']??$payload['user_id']??'';$room=$space['name']??$payload['space_name']??'';if($user===''||$room==='')throw new InvalidArgumentException('Google Chat event has no sender or space');$attachment=$message['attachments'][0]??$message['attachment'][0]??null;return ['platform'=>'google_chat','user_id'=>(string)$user,'reply_target'=>(string)$room,'content_type'=>$attachment?'file':'text','text'=>$message['text']??null,'media_url'=>$attachment['downloadUri']??$attachment['resourceName']??null];
}
function google_chat_send_reply(array $event,array $reply,string $token):void
{
    if(empty($event['reply_target'])||$token==='')throw new InvalidArgumentException('Google Chat space and token are required');foreach(array_slice($reply['messages']??[],0,5)as$message){$ch=adapter_curl_init('https://chat.googleapis.com/v1/'.$event['reply_target'].'/messages');curl_setopt_array($ch,[CURLOPT_POST=>true,CURLOPT_RETURNTRANSFER=>true,CURLOPT_HTTPHEADER=>['Authorization: Bearer '.$token,'Content-Type: application/json'],CURLOPT_POSTFIELDS=>json_encode(['text'=>$message['text']??$message['media_url']??''],JSON_THROW_ON_ERROR)]);$result=curl_exec($ch);$status=curl_getinfo($ch,CURLINFO_RESPONSE_CODE);if($result===false||$status<200||$status>=300)throw new RuntimeException('Google Chat API request failed');curl_close($ch);}
}

File: python/platforms/google_chat.py
from __future__ import annotations
import os
from typing import Any
import requests
from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG

class GoogleChatAdapter(PlatformAdapter):
    """Google Chat app adapter using spaces.messages.create."""
    def __init__(self, token: str | None = None, session: requests.Session | None = None):
        self.capabilities = PLATFORM_CATALOG["google_chat"]
        self.token = token or os.getenv("GOOGLE_CHAT_ACCESS_TOKEN")
        if not self.token: raise ValueError("GOOGLE_CHAT_ACCESS_TOKEN is required")
        self.session = session or requests.Session()
    def parse_event(self, payload: Any, headers=None) -> InboundEvent:
        message = payload.get("message", payload); sender = message.get("sender", {})
        space = payload.get("space", {})
        attachments = message.get("attachment") or message.get("attachments") or []
        attachment = attachments[0] if attachments else {}
        media = attachment.get("downloadUri") or attachment.get("resourceName")
        user_id = sender.get("name") or payload.get("user_id")
        space_name = space.get("name") or payload.get("space_name")
        if not user_id or not space_name:
            raise ValueError("Google Chat event has no sender or space")
        return InboundEvent(platform="google_chat", user_id=str(user_id), reply_target=str(space_name), content_type="file" if media else "text", text=message.get("text"), media_url=media)
    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        space = event.reply_target or os.getenv("GOOGLE_CHAT_SPACE")
        if not space: raise ValueError("Google Chat space is required")
        for message in reply.messages:
            response = self.session.post(f"https://chat.googleapis.com/v1/{space}/messages", headers={"Authorization": f"Bearer {self.token}"}, json={"text": message.text or message.media_url or ""}, timeout=30)
            response.raise_for_status()
